File: packages/js/core/src/models/dit.ts
import { Tensor } from '../ir/tensor.js';
import { Gelu, Gemm, LayerNormalization, MultiHeadAttention } from '../primitives.js';
import { PatchEmbed } from './vit.js';

function getParam(name: string, shape: number[], dtype: any = 'float32'): Tensor {
  return new Tensor(name, shape, dtype, false, false, new Float32Array());
}

function recordOp(opType: string, inputs: Tensor[], attr?: any): Tensor {
  const dtype = inputs[0]?.dtype ?? 'float32';
  return new Tensor(`${opType}_out`, [], dtype, false, false, new Float32Array());
}

export class DiTBlock {
  public prefix: string;
  public hiddenSize: number;
  public norm1: LayerNormalization;
  public attn: MultiHeadAttention;
  public norm2: LayerNormalization;
  public mlpFc1: Gemm;
  public act: Gelu;
  public mlpFc2: Gemm;
  public adaLNModulation: Gemm;

  constructor(hiddenSize: number, numHeads: number, mlpRatio: number = 4.0, prefix: string = '') {
    this.prefix = prefix;
    this.hiddenSize = hiddenSize;
    this.norm1 = new LayerNormalization([hiddenSize], 1e-6);
    this.attn = new MultiHeadAttention(numHeads, true, true);
    this.norm2 = new LayerNormalization([hiddenSize], 1e-6);

    this.mlpFc1 = new Gemm(1.0, 1.0, 0, 1);
    this.act = new Gelu();
    this.mlpFc2 = new Gemm(1.0, 1.0, 0, 1);
    this.adaLNModulation = new Gemm(1.0, 1.0, 0, 1);
  }

  call(x: Tensor, c: Tensor): Tensor {
    const cProj = this.adaLNModulation.call(
      c,
      getParam(`${this.prefix}.adaLN_modulation.weight`, [6 * this.hiddenSize, this.hiddenSize]),
      getParam(`${this.prefix}.adaLN_modulation.bias`, [6 * this.hiddenSize]),
    );

    let splitOut = recordOp('Split', [cProj]);

    const axisTensor = recordOp('Constant', [], { value: [1], dtype: 7 });
    let shiftMsa = recordOp('Unsqueeze', [splitOut, axisTensor]);
    let scaleMsa = recordOp('Unsqueeze', [splitOut, axisTensor]);
    let gateMsa = recordOp('Unsqueeze', [splitOut, axisTensor]);
    let shiftMlp = recordOp('Unsqueeze', [splitOut, axisTensor]);
    let scaleMlp = recordOp('Unsqueeze', [splitOut, axisTensor]);
    let gateMlp = recordOp('Unsqueeze', [splitOut, axisTensor]);

    let identity = x;
    let xNorm = this.norm1.call(
      x,
      getParam(`${this.prefix}.norm1.weight`, [this.hiddenSize]),
      getParam(`${this.prefix}.norm1.bias`, [this.hiddenSize]),
    );

    const oneTensor = recordOp('Constant', [], { value: [1.0], dtype: 1 });
    let xModulated = recordOp('Add', [
      recordOp('Mul', [recordOp('Add', [oneTensor, scaleMsa]), xNorm]),
      shiftMsa,
    ]);

    let xAttn = this.attn.call(xModulated, xModulated, xModulated);
    xAttn = recordOp('Mul', [gateMsa, xAttn]);
    x = recordOp('Add', [identity, xAttn]);

    identity = x;
    xNorm = this.norm2.call(
      x,
      getParam(`${this.prefix}.norm2.weight`, [this.hiddenSize]),
      getParam(`${this.prefix}.norm2.bias`, [this.hiddenSize]),
    );

    xModulated = recordOp('Add', [
      recordOp('Mul', [recordOp('Add', [oneTensor, scaleMlp]), xNorm]),
      shiftMlp,
    ]);

    let xMlp = this.mlpFc1.call(
      xModulated,
      getParam(`${this.prefix}.mlp.fc1.weight`, [this.hiddenSize * 4, this.hiddenSize]),
      getParam(`${this.prefix}.mlp.fc1.bias`, [this.hiddenSize * 4]),
    );
    xMlp = this.act.call(xMlp);
    xMlp = this.mlpFc2.call(
      xMlp,
      getParam(`${this.prefix}.mlp.fc2.weight`, [this.hiddenSize, this.hiddenSize * 4]),
      getParam(`${this.prefix}.mlp.fc2.bias`, [this.hiddenSize]),
    );

    xMlp = recordOp('Mul', [gateMlp, xMlp]);
    x = recordOp('Add', [identity, xMlp]);

    return x;
  }
}

export class DiT {
  public hiddenSize: number;
  public outChannels: number;
  public patchEmbed: PatchEmbed;
  public tEmbedder: Gemm;
  public blocks: DiTBlock[];
  public finalLayerNorm: LayerNormalization;
  public finalLayerAdaLN: Gemm;
  public finalLayerProj: Gemm;

  constructor(
    inputSize: number = 32,
    patchSize: number = 2,
    inChannels: number = 4,
    hiddenSize: number = 1152,
    depth: number = 28,
    numHeads: number = 16,
    mlpRatio: number = 4.0,
  ) {
    this.hiddenSize = hiddenSize;
    this.outChannels = inChannels * patchSize * patchSize;
    this.patchEmbed = new PatchEmbed(inputSize, patchSize, inChannels, hiddenSize, 'x_embedder');
    this.tEmbedder = new Gemm(1.0, 1.0, 0, 1);

    this.blocks = [];
    for (let i = 0; i < depth; i++) {
      this.blocks.push(new DiTBlock(hiddenSize, numHeads, mlpRatio, `blocks.${i}`));
    }

    this.finalLayerNorm = new LayerNormalization([hiddenSize], 1e-6);
    this.finalLayerAdaLN = new Gemm(1.0, 1.0, 0, 1);
    this.finalLayerProj = new Gemm(1.0, 1.0, 0, 1);
  }

  call(x: Tensor, t: Tensor): Tensor {
    x = this.patchEmbed.call(x);

    const posEmbed = getParam('pos_embed', [1, this.patchEmbed.numPatches, this.hiddenSize]);
    x = recordOp('Add', [x, posEmbed]);

    const c = this.tEmbedder.call(
      t,
      getParam('t_embedder.weight', [this.hiddenSize, this.hiddenSize]),
      getParam('t_embedder.bias', [this.hiddenSize]),
    );

    for (const block of this.blocks) {
      x = block.call(x, c);
    }

    const cProj = this.finalLayerAdaLN.call(
      c,
      getParam('final_layer.adaLN_modulation.weight', [2 * this.hiddenSize, this.hiddenSize]),
      getParam('final_layer.adaLN_modulation.bias', [2 * this.hiddenSize]),
    );

    let splitOut = recordOp('Split', [cProj]);
    const axisTensor = recordOp('Constant', [], { value: [1], dtype: 7 });
    let shift = recordOp('Unsqueeze', [splitOut, axisTensor]);
    let scale = recordOp('Unsqueeze', [splitOut, axisTensor]);

    x = this.finalLayerNorm.call(
      x,
      getParam('final_layer.norm.weight', [this.hiddenSize]),
      getParam('final_layer.norm.bias', [this.hiddenSize]),
    );
    const oneTensor = recordOp('Constant', [], { value: [1.0], dtype: 1 });
    x = recordOp('Add', [recordOp('Mul', [recordOp('Add', [oneTensor, scale]), x]), shift]);

    x = this.finalLayerProj.call(
      x,
      getParam('final_layer.linear.weight', [this.outChannels, this.hiddenSize]),
      getParam('final_layer.linear.bias', [this.outChannels]),
    );
    return x;
  }
}

export function ditXl2(): DiT {
  return new DiT(32, 2, 4, 1152, 28, 16, 4.0);
}

File: packages/python/onnx9000-core/src/onnx9000/core/models/dit.py
"""Dit."""

from typing import Any

from onnx9000.core.ir import Tensor, Variable
from onnx9000.core.models.vit import PatchEmbed
from onnx9000.core.ops import add, mul
from onnx9000.core.primitives import Gelu, Gemm, LayerNormalization, MultiHeadAttention


def get_param(name: str, shape: list[int], dtype: int = 1) -> Variable:  # noqa: D103
    """Get param."""
    return Variable(name=name, shape=shape, dtype=dtype)


class DiTBlock:
    """Diffusion Transformer Block with AdaLN-Zero."""

    def __init__(self, hidden_size: int, num_heads: int, mlp_ratio: float = 4.0, prefix: str = ""):  # noqa: D107
        """Init."""
        self.prefix = prefix
        self.hidden_size = hidden_size
        self.norm1 = LayerNormalization((hidden_size,), epsilon=1e-6)
        self.attn = MultiHeadAttention(num_heads=num_heads, qkv_bias=True)
        self.norm2 = LayerNormalization((hidden_size,), epsilon=1e-6)
        int(hidden_size * mlp_ratio)
        self.mlp_fc1 = Gemm(trans_b=1)
        self.act = Gelu()
        self.mlp_fc2 = Gemm(trans_b=1)
        self.adaLN_modulation = Gemm(trans_b=1)  # projects time_emb to shift, scale, gate

    def __call__(self, x: Tensor, c: Tensor) -> Tensor:  # noqa: D102
        # c is the timestep/condition embedding [B, hidden_size]
        """Call."""
        from onnx9000.core.ops import split, squeeze

        # AdaLN-Zero mapping
        # We need 6 chunks: shift_msa, scale_msa, gate_msa, shift_mlp, scale_mlp, gate_mlp
        c_proj = self.adaLN_modulation(
            c,
            get_param(
                f"{self.prefix}.adaLN_modulation.weight", [6 * self.hidden_size, self.hidden_size]
            ),
            get_param(f"{self.prefix}.adaLN_modulation.bias", [6 * self.hidden_size]),
        )

        # For AST builder mock, split returns a single tensor, so we just duplicate it
        split_out = split(c_proj)
        chunks = [split_out] * 6
        shift_msa, scale_msa, gate_msa, shift_mlp, scale_mlp, gate_mlp = chunks

        # Needs unsqueeze to broadcast across sequence length
        from onnx9000.core.ops import record_op, unsqueeze

        axis_tensor = record_op("Constant", [], {"value": [1], "dtype": 7})
        shift_msa = record_op("Unsqueeze", [shift_msa, axis_tensor])
        scale_msa = record_op("Unsqueeze", [scale_msa, axis_tensor])
        gate_msa = record_op("Unsqueeze", [gate_msa, axis_tensor])
        shift_mlp = record_op("Unsqueeze", [shift_mlp, axis_tensor])
        scale_mlp = record_op("Unsqueeze", [scale_mlp, axis_tensor])
        gate_mlp = record_op("Unsqueeze", [gate_mlp, axis_tensor])

        # Apply norm and modulation for attention
        identity = x
        x_norm = self.norm1(
            x,
            get_param(f"{self.prefix}.norm1.weight", [self.hidden_size]),
            get_param(f"{self.prefix}.norm1.bias", [self.hidden_size]),
        )

        # (1 + scale_msa) * x_norm + shift_msa
        one_tensor = record_op("Constant", [], {"value": [1.0], "dtype": 1})
        x_modulated = add(mul(add(one_tensor, scale_msa), x_norm), shift_msa)

        x_attn = self.attn(x_modulated, x_modulated, x_modulated)

        # gate_msa * x_attn
        x_attn = mul(gate_msa, x_attn)
        x = add(identity, x_attn)

        # Apply norm and modulation for MLP
        identity = x
        x_norm = self.norm2(
            x,
            get_param(f"{self.prefix}.norm2.weight", [self.hidden_size]),
            get_param(f"{self.prefix}.norm2.bias", [self.hidden_size]),
        )

        x_modulated = add(mul(add(one_tensor, scale_mlp), x_norm), shift_mlp)

        x_mlp = self.mlp_fc1(
            x_modulated,
            get_param(
                f"{self.prefix}.mlp.fc1.weight", [int(self.hidden_size * 4), self.hidden_size]
            ),
            get_param(f"{self.prefix}.mlp.fc1.bias", [int(self.hidden_size * 4)]),
        )
        x_mlp = self.act(x_mlp)
        x_mlp = self.mlp_fc2(
            x_mlp,
            get_param(
                f"{self.prefix}.mlp.fc2.weight", [self.hidden_size, int(self.hidden_size * 4)]
            ),
            get_param(f"{self.prefix}.mlp.fc2.bias", [self.hidden_size]),
        )

        x_mlp = mul(gate_mlp, x_mlp)
        x = add(identity, x_mlp)

        return x


class DiT:  # noqa: D101
    """Di t."""

    def __init__(  # noqa: D107
        self,
        input_size: int = 32,
        patch_size: int = 2,
        in_channels: int = 4,
        hidden_size: int = 1152,
        depth: int = 28,
        num_heads: int = 16,
        mlp_ratio: float = 4.0,
    ):
        """Init."""
        self.hidden_size = hidden_size
        self.patch_embed = PatchEmbed(
            input_size, patch_size, in_channels, hidden_size, prefix="x_embedder"
        )
        self.t_embedder = Gemm(trans_b=1)  # Time embedding

        self.blocks = [
            DiTBlock(hidden_size, num_heads, mlp_ratio, prefix=f"blocks.{i}") for i in range(depth)
        ]

        self.final_layer_norm = LayerNormalization((hidden_size,), epsilon=1e-6)
        self.final_layer_adaLN = Gemm(trans_b=1)
        self.final_layer_proj = Gemm(trans_b=1)

        self.out_channels = in_channels * patch_size * patch_size

    def __call__(self, x: Tensor, t: Tensor) -> Tensor:  # noqa: D102
        """Call."""
        x = self.patch_embed(x)

        pos_embed = get_param("pos_embed", [1, self.patch_embed.num_patches, self.hidden_size])
        x = add(x, pos_embed)

        c = self.t_embedder(
            t,
            get_param("t_embedder.weight", [self.hidden_size, self.hidden_size]),
            get_param("t_embedder.bias", [self.hidden_size]),
        )

        for block in self.blocks:
            x = block(x, c)

        # Final AdaLN
        from onnx9000.core.ops import record_op, split

        c_proj = self.final_layer_adaLN(
            c,
            get_param(
                "final_layer.adaLN_modulation.weight", [2 * self.hidden_size, self.hidden_size]
            ),
            get_param("final_layer.adaLN_modulation.bias", [2 * self.hidden_size]),
        )

        split_out = split(c_proj)
        shift, scale = split_out, split_out
        axis_tensor = record_op("Constant", [], {"value": [1], "dtype": 7})
        shift = record_op("Unsqueeze", [shift, axis_tensor])
        scale = record_op("Unsqueeze", [scale, axis_tensor])

        x = self.final_layer_norm(
            x,
            get_param("final_layer.norm.weight", [self.hidden_size]),
            get_param("final_layer.norm.bias", [self.hidden_size]),
        )
        one_tensor = record_op("Constant", [], {"value": [1.0], "dtype": 1})
        x = add(mul(add(one_tensor, scale), x), shift)

        x = self.final_layer_proj(
            x,
            get_param("final_layer.linear.weight", [self.out_channels, self.hidden_size]),
            get_param("final_layer.linear.bias", [self.out_channels]),
        )
        return x


def dit_xl_2(**kwargs: Any) -> DiT:  # noqa: D103
    """Dit xl 2."""
    return DiT(hidden_size=1152, depth=28, num_heads=16, **kwargs)
